[ENH] Preprocessor widget: add random feature selection

matevzkren opened the pull request:
Added a new preprocessor to the Preprocessor widget. Selects
features at random. Both selection per proportion and selection per number of
features implemented. FIxes #1087.

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +535,112 @@ def createinstance(params):
+
+        # form.addRow(QRadioButton("FDR"), QDoubleSpinBox())
+        # form.addRow(QRadioButton("FPR"), QDoubleSpinBox())
+        # form.addRow(QRadioButton("FWE"), QDoubleSpinBox())
```
Commented-out lines in the code are only allowed if preceded by a comment explaining why they are there.

kernc commented on Orange/widgets/data/icons/SelectColumnsRandom.svg:
```diff
@@ -0,0 +1,39 @@
+		c-1.116,0-2.032,0.329-2.748,0.987C20.563,14.898,20.2,15.746,20.188,16.784z M23.664,30.925c0.885,0,1.641,0.304,2.266,0.91
+		c0.625,0.607,0.938,1.341,0.938,2.202c0,0.862-0.311,1.599-0.928,2.212c-0.619,0.612-1.365,0.919-2.24,0.919
+		c-0.873,0-1.616-0.307-2.229-0.919c-0.613-0.613-0.919-1.35-0.919-2.212c0-0.861,0.303-1.595,0.91-2.202
+		C22.068,31.229,22.802,30.925,23.664,30.925z"/>
```
Could the white border around the question mark be made a pixel or two wider?

![screenshot - 03122016 - 08 51 51 pm](https://cloud.githubusercontent.com/assets/684364/13724910/8375fef8-e894-11e5-943f-4581dd46f05d.png)

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +535,112 @@ def createinstance(params):
+        self.__spins[RandomFeatureSelectEditor.Fixed] = kspin
+        form.addRow(fixedrb, kspin)
+
+        percrb = QRadioButton("Percentile")
```
I think ["percentile" should really be "percentage"](https://www.google.com/search?q=percent+vs+percentile) in this context. @janezd will confirm.

matevzkren replied:
On that note, do you prefer 'Random Feature Selection' or 'Select Random Features' for the description title?

kernc replied:
Indeed. The latter one matches the line above it as well as grammatically all the rest that start with a verb.

matevzkren commented:
I've corrected the icon and implemented the requested changes. Percentile -> percentage swap was also pushed preemptively, but can be quickly reverted at any point if decided against.

lanzagar commented:
Bug 1:
File (wine) -> Preprocess (Select Random Features)
Then change to iris dataset in File.
It still tries to sample 10 features which results in an error.

Bug 2:
Check the Auto Commit option in Preprocess, change settings for Select Random Features. They are not commited.

matevzkren commented:
@lanzagar Both preprocessors (Select Random & Select Relevat) auto-commit alike when spinBox = deselected && value = changed. What did you have in mind?

lanzagar commented:
I didn't notice it updated on editingFinished... I expected to see changes happen immediately when using the spin controls to increase/decrease the value.
We might want to reconsider this in general (for Select Relevant and in other places as well).

As for the number of features - it now does not crash on numbers higher then number of features anymore, but they can still be selected. I think the maximum value in the spin box should be in sync with the data size.

Also @kernc mentioned you might want to use the spin control from Orange.widgets.gui...

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +535,108 @@ def createinstance(params):
+            enabled=self.__strategy == RandomFeatureSelectEditor.Fixed
+        )
+        kspin.valueChanged[int].connect(self.setK)
+        kspin.editingFinished.connect(self.edited)
```
To simplify, can you perhaps use `gui.spin()`?

kernc commented on Orange/preprocess/fss.py:
```diff
@@ -121,7 +121,8 @@ def __call__(self, data):
         domain = Orange.data.Domain(
-            random.sample(data.domain.attributes, self.k),
+            random.sample(data.domain.attributes,
+                          min(self.k, len(data.domain.attributes))),
```
Is it possible to perhaps set this maximum also on the UI spin box?

matevzkren commented:
> I didn't notice it updated on editingFinished... I expected to see changes happen immediately when using the spin controls to increase/decrease the value.
> We might want to reconsider this in general (for Select Relevant and in other places as well).

@lanzagar Though random feature selection is extremely lightweight, I don't think an update on every spin click _in general_ is preferred. With calculation time usually dependant on data size, the current update practice feels safer. Admittedly, there might be valid reasoning I'm unaware of!

kernc commented:
@matevzkren, what do you think, could you add checks so that no more than `len(data.domain.attributes)` features can be selected in spin boxes and that these maximums update when new data is connected?

Regarding `gui.spin`, I'm not as certain.

matevzkren commented:
@kernc Please consider this to be a work in progress. I wanted to play around with gui.py and see what could be created. I'm starting to realise you only intended for me to use `gui.spin` however. Note that everything can be changed (depending on what you think is better, I **can avoid gui.py** altogether or only use `gui.spin`).

Several extras were implemented, including how percentage `gui.hSlider` is presented and handled (`QAbstractSlider.setTracking` simply didn't meet my expectations). Underlying spins are disabled when a different strategy is in play. No more than `len(data.domain.attributes)` features can even be selected.

Overall, the code is somewhat specific to this preprocessor. For example, `k_spin` is inserted into the radio button. If required, the implementation can be simplified and some extras taken away.

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +536,125 @@ def createinstance(params):
+        prev_val = self._ticks_to_update_slider[1]
+        if (self._ticks_to_update_slider[0] is not None and (prev_val is None or
+            any([curr_val < tick <= prev_val or curr_val >= tick > prev_val
+                for tick in self._ticks_to_update_slider[0]]))):
```
[Pylint reports](https://travis-ci.org/biolab/orange3/jobs/120086851) bad-continuation for this condition. Please make it neater.

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +536,125 @@ def createinstance(params):
+        curr_val = self._p
+        prev_val = self._ticks_to_update_slider[1]
+        if (self._ticks_to_update_slider[0] is not None and (prev_val is None or
+            any([curr_val < tick <= prev_val or curr_val >= tick > prev_val
```
`any()` is great, but it also accepts a generator and returns as soon as it finds the first truish item. No need to build the whole list.

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +536,125 @@ def createinstance(params):
+                p_spin.setTickInterval(p_spin.maximum() /
+                                       self._max_num_attributes)
+                self._ticks_to_update_slider = \
+                    [[math.ceil(i*p_spin.maximum()/self._max_num_attributes)
```
`numpy` is already imported and it provides `ceil()`.

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +536,125 @@ def createinstance(params):
 
+class RandomFeatureSelectEditor(BaseEditor):
+    #: Strategy
+    Fixed, Percentage, FDR, FPR, FWE = 0, 1, 2, 3, 4
```
What are FDR, FPR, and FWE?

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +536,125 @@ def createinstance(params):
+            widget=ibox, master=self, value="_p", minValue=0, maxValue=100,
+            step=4, ticks=20, labelFormat="%d %%", callback=self.update_p_spin,
+            disabled=self._strategy != RandomFeatureSelectEditor.Percentage)
+        self._spins[RandomFeatureSelectEditor.Percentage] = p_spin
```
Percentage editor is not a spin box. You don't even need a reference to "`p_spin`" and if you use gui.py constructs, you don't need `set_p()` method — you simply assign to `self._p` and the slider's value will be updated. Similar goes for `self._k` and its spin box, except you need its reference for setting the max value.

ales-erjavec replied:
> ... gui.py constructs, you don't need set_p() method — you simply assign to self._p and the slider's value will be updated.

I doubt that would work in this case.

kernc replied:
Oh, right, that only works in classes that extend `OWComponent`. :unamused:

kernc replied:
@matevzkren, with this not working out of the box, I don't think there is much reason to obfuscate code by using gui.py widgets. So either have this class extend `gui.OWComponent` (some adjustment might be necessary) to reap its benefits (i.e. dropping the setter methods as noted in the comment above), or revert to using Qt directly as the other widgets here do (in which case I'm sorry to have misled you :smile:).

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +536,125 @@ def createinstance(params):
+            else:
+                k_spin.setMaximum(self._max_num_attributes)
+                p_spin.setTickInterval(p_spin.maximum() /
+                                       self._max_num_attributes)
```
No need to adjust `p_spin`. It has percentage granularity since initialization, and that's that.

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -1803,6 +1938,8 @@ def __on_modelchanged(self):
         """Set the input data set."""
         self.data = data
+        self.controler.set_max_num_attributes(len(data.domain.attributes) if
+                                              self.data else None)
```
Bad line breaking. Before `if` would be better, or the whole argument in another line. Please don't rely on PyCharm doing too much of your work. :laughing:

kernc commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -535,6 +536,125 @@ def createinstance(params):
+        self.set_p(params.get("p", 75))
+        self.set_max_num_attributes(params.get("max_num_attributes", None))
+        self.set_strategy(params.get("strategy",
+                                     RandomFeatureSelectEditor.Fixed))
```
You can just assign to `self._k`, `self._p`, and `self._strategy` here, and their respective controls will be updated. That's what `master, value` parameters of gui.py controls are for. Then you probably also don't need `set_strategy()` and `update_strategy()` methods. If `parameters()` and `createinstance()` below are the only two entry points for `OWPreprocess` to use, you don't even need callbacks on your gui.py elements.

kernc commented:
I left you some comments. Looks good and seems to work well. Perhaps the previous commiter (@ales-erjavec) would prefer you use his provided `edited, changed, setParameters()` interface. I'm not so sure mixing the styles is such a good idea (:laughing:), but I like gui.py. As you can see from my comments above, it makes Orange GUI programming somewhat easier, most of the time.

ales-erjavec commented on Orange/widgets/data/owpreprocess.py:
```diff
@@ -1258,6 +1386,13 @@ def setModelData(self, widget, index):
+            row = self.view.indexOf(widget)
+            index = self.model().index(row, 0)
+            self.setWidgetData(widget, index)
+
```
This should not be here (in this class).

In fact the individual preprocessors (edited by this widget) should not (in fact can't) be dependent on the input data.

matevzkren commented:
Thanks for the comments,

@kernc I chose to reference "p_spin" for the following reasons. `hSliders` do have percentage granularity enabled by default, but it remains the same throughout, as far as I can tell. I chose to edit it in an attempt to offer the user a visual connection, i.e. portray how many variables there are in total and how many more percent to trigger the next threshold. If the feature isn't needed, I'll delete. Writing it down as a spin was to simplify, but I can change that.

Also, the random selection of features is not ran on every percentage tick. `self.edited.emit()` is only used when the conditions are met and a different integer number of variables is requested. This causes a lot less updates and feels nicer overall (`update_p_spin`).

In regard to the auxiliary functions, I primarily wanted to match the signals from previous implementations of other preprocessors. Imo, `update_strategy()`is still needed to set the correct `setDisabled` marks (that seems to not happen automatically, but I could be missing something). Having removed the auxilliary functions, the preprocessor worked fine, but was definitely not as fluid overall. I will look into it more tomorrow, then report back.

@ales-erjavec I made some changes, but I'm not sure whether it's correct. Either way, some further explanation would be appreciated (if you don't mind).

Thanks to both, again. I'm not sure how I'm letting some of these mistakes through, but hey-yo ^^

kernc commented:
> I chose to reference "p_spin" for the following reasons. `hSliders` do have percentage granularity enabled by default, but it remains the same throughout, as far as I can tell. I chose to edit it in an attempt to offer the user a visual connection, i.e. portray how many variables there are in total and how many more percent to trigger the next threshold. If the feature isn't needed, I'll delete. Writing it down as a spin was to simplify, but I can change that.

> Also, the random selection of features is not ran on every percentage tick. `self.edited.emit()` is only used when the conditions are met and a different integer number of variables is requested. This causes a lot less updates and feels nicer overall (`update_p_spin`).

Understood. IMHO `data[:, selected_cols]` was never a bottleneck and we should dedicate exactly 0% of code complexity to such optimizations.

ales-erjavec commented:
> @ales-erjavec I made some changes, but I'm not sure whether I'm in the clear. If you don't mind, some further explanation would be appreciated.

Regarding the max_num_attributes. When the preprocessor will actually get the data it will in general not have the same number of attributes as input data to the widget, but the result of applying the preprocessors (on the input data) before it. In particular that number could be larger.
Example: Imagine a dataset with 2 discrete features each with 5 values/levels and the first preprocessor applied is a "Continuize" which expands the domain into 10 indicator columns. A "Random Feature Selection" after it would still only allow the user to select 2 features. I think that having the GUI impose such arbitrary limits is counter productive.
